### PR TITLE
Preserve scope for insert_all and friends.

### PR DESCRIPTION
**Pull request by simi**
Alternative approach to https://github.com/rails/rails/pull/35626 based on https://github.com/rails/rails/pull/35635 using `scope_for_create` to assign relation defaults reverse merged with passed defaults as an argument => relation scope is used only if the same key is not present in defaults or in data hash.

It is implemented directly on relation = all associations should be supported as well.

```ruby
Author.first.books.where(published: true).insert_all([{ name: "My little book", isbn: "197452259"}])
```

Creates one book with assigned author and `published: true` taken from where relation.

If approved, I can add all friend methods (`upsert_all`, `insert_all!`, `insert`, `upsert`, `insert!`) and add some documentations as well.

### CODE REVIEW

**Comment by simi** [comment]
@kamipo per https://github.com/rails/rails/pull/35626#issuecomment-473497384 `test_has_many_through_create` fails. Anyway `insert_all` on relation works with this PR.

**Review by boblail — CHANGES_REQUESTED** [review]
❗️ There's a difference in behavior between this PR and #35626 that's worth a discussion!

This PR will `reverse_merge` implicit scope into inserted values whereas #35626 will `merge`.

Given this example,
> ```ruby
> Author.find(5).books.insert_all([
>   { title: "Rework", author_id: 7 }
> ])
> ```

1. on `abhaynikam:35493-insert-all-relation`, the new book will belong to Author 5
2. on `simi:insert_all_relation`, the new book will belong to Author 7

👉 I believe we should close _this_ PR in favor of #35626 because `create` behaves the first way:
> ```ruby
> Author.find(5).books.create([
>   { title: "Rework", author_id: 7 }
> ])
> # the new book belongs to Author 5
> ```

**Comment by simi** [comment]
@boblail that's something I can easily change actually. There's no need to close this PR just because this. I have opened this PR to introduce different approach on internals (reusing `defaults` option, use `scope_for_create`, introduce it directly on `Relation` class so it works on `where` and on relations as well, ...).

**Comment by boblail** [comment]
@simi Ah, fair enough! I guess I was assuming that implementing in terms of `defaults` was fundamental to this PR, but obviously not! 😄 

(and I think implementing this feature in terms of `defaults` would result in the wrong behavior here)

**Comment by simi** [comment]
@boblail There separate PR for `defaults` functionality - https://github.com/rails/rails/pull/35635.

**Comment by simi** [comment]
If the approach should be the same as `create` method, I have different idea to pass `scope_for_create` to `InsertAll` making it preferred source of data. The lookup will be then `scope_for_create`, `insert_values`, `defaults` on `Relation`. It will be used only when called from `Relation`.

WDYT @boblail?

**Comment by boblail** [comment]
@simi, I'm good with that! 👍 

I like how, with `:defaults`, you chained the lookups rather than looping through `attributes`.

— and I like the idea of using `scope_for_create`.

I wonder if we really need to pass `scope_for_create` to `InsertAll` as its own argument, though... Like: `InsertAll`'s first argument is `model`... Could that be `scope` instead? — it doesn't even have to be renamed as long as it's a proxy for model. (Does `Persistence` have access to current scope?)

**Comment by simi** [comment]
@boblail I have pushed changes to pass `context` to `InsertAll` and try to use `scope_for_create` if possible as preferred way to lookup for attributes.

File: activerecord/test/cases/insert_all_test.rb
Comment (simi) [inline activerecord/test/cases/insert_all_test.rb]: :thinking: Those specs can be rewritten to use `insert_all!` later once approach will be accepted.

**Comment by simi** [comment]
@boblail I have found a way how to pass scope from `Relation` and from `Persistence` as well. See my latest change.

**Comment by simi** [comment]
I have spent some more time on this and I have found out (obviously) "through" relations doesn't work. I think it is possible to support them as well, but to keep this pull request small as possible, it can just raise when insert_all is called on through association for now.

To support through associations I think we will need to begin transaction, insert_all at target table, return ids and compose insert data for second insert_all to through association. We can reuse scope_for_context of through_association (https://github.com/rails/rails/blob/d8d6bd5e63a9a4a6c06a4dde3c7137ee2be105fd/activerecord/lib/active_record/associations/has_many_through_association.rb#L60) and through_scope_attributes to compose insert data. It will result into two or more insert_all calls in transaction. But should be still huge improvement. I'm just a little afraid of all edge cases code is already handling.

**Review by kaspth — COMMENTED** [review]
Heads up: while reviewing this I wanted to simplify the code in `values_list` some more. Now builder shouldn't have to know about either scope_for_create/attributes/defaults because of map_key_with_value.

**Comment by kaspth** [comment]
`insert_all` has turned in to quite greedy for config options. So I'm wondering if we should just cut `:defaults` and just support the relation aspects.

File: activerecord/lib/active_record/insert_all.rb
Comment (kaspth) [inline activerecord/lib/active_record/insert_all.rb]: @kamipo would it make sense to extract the below from here into something like `bind_value_for_database`?

```ruby
bind = Relation::QueryAttribute.new(key, attributes[key], types[key])
connection.with_yaml_fallback(bind.value_for_database)
```

InsertAll has a fairly intimate relationship with connection and also sort of Relation::QueryAttribute. I guess the tough part is `types[key]` because we might not want to run `lookup_cast_type_from_column` on the M*N loop.

**Comment by simi** [comment]
@kaspth defaults are separated feature. First implementation of this relied on defaults, but now they are not needed, I'll rebase it out of this pull request.

There is separate PR only for defaults functionality - https://github.com/rails/rails/pull/35635.

File: activerecord/lib/active_record/relation.rb
Comment (simi) [reply activerecord/lib/active_record/relation.rb]: It will lose the `scope_for_context` on relation then.

File: activerecord/lib/active_record/insert_all.rb
Comment (simi) [reply activerecord/lib/active_record/insert_all.rb]: `scope_for_create` is empty or has only scope attributes (for example `{"author_id" => 5}`.
`defaults` are additional attributes passed to not repeat the same attribute in each payload array of hashes called `inserts` in here. We need to merge all of those to get all final keys.

File: activerecord/lib/active_record/insert_all.rb
Comment (simi) [reply activerecord/lib/active_record/insert_all.rb]: Good idea. Anyway I have removed `defaults` from this PR, but I'll definitely keep this in my mind for later. :+1:

**Comment by simi** [comment]
I have rebased this to reflect latest changes from `master` on `InsertAll` class. `Builder` doesn't need to know `scope_for_create` now. :+1:

File: activerecord/lib/active_record/relation.rb
Comment (boblail) [inline activerecord/lib/active_record/relation.rb]: After realizing you could pass `scope_for_association` to `InsertAll` from `Persistence` (👍), can we drop this method from `Relation`?

File: activerecord/lib/active_record/insert_all.rb
Comment (boblail) [inline activerecord/lib/active_record/insert_all.rb]: Can just be `attributes.stringify_keys.merge(scope_for_create)` instead of `reverse_merge` — it's a little more straightforward that way.

File: activerecord/test/cases/insert_all_test.rb
Comment (boblail) [inline activerecord/test/cases/insert_all_test.rb]: Could you add a test (or tests) to cover also `create_with`, which this PR adds support for? — loosely like:

```ruby
Book.create_with(isbn: "X").insert_all([ { name: "A" }, { name: "B" } ])
assert_equal 2, Book.where(isbn: "X").count
```

**Review by boblail — COMMENTED** [review]
👍 This looks great to me, @simi! Thanks for doing this!!

**Comment by simi** [comment]
Thanks for another review round @boblail. I'm waiting for final decision in https://github.com/rails/rails/pull/35686 to continue work on all my pull requests related to `insert_all`.

**Comment by simi** [comment]
Since #35686 was closed I'm going to continue on this.

My plan is to raise on "through" relation for now. I can open another PR later to support this as well.

**Comment by simi** [comment]
Before I'll do (:pray:) final take on this, I would like to ensure what is needed to be done.

1. I'll go thru @boblail comments and do code changes
2. I'll make through relation to raise for now. I can open my idea how to implement it in there as well, it still should be possible.

In another PR we can also improve rails guides. Or we can also decide to not promote it for now.

Is there anything else needed in here?

File: activerecord/test/cases/insert_all_test.rb
Comment (simi) [reply activerecord/test/cases/insert_all_test.rb]: done

File: activerecord/lib/active_record/insert_all.rb
Comment (simi) [inline activerecord/lib/active_record/insert_all.rb]: It is not the best way how to check for `through` relation, but I wasn't able to find better for now.

File: activerecord/lib/active_record/persistence.rb
Comment (simi) [inline activerecord/lib/active_record/persistence.rb]: We need to pass `current_scope` and fallback to `relation` (default argument) for now. `scope_for_association` with default argument doesn't preserve `scope_for_create`. Isn't that actually bug? I would expect `scope_for_association` to keep `scope_for_create`. I can change `scope_for_association` to keep that.

File: activerecord/lib/active_record/relation.rb
Comment (simi) [reply activerecord/lib/active_record/relation.rb]: fixed, there's no need to redefine `insert_all` for relation now, but it will needs still few tweaks, see my latest comments.

**Comment by simi** [comment]
> 1. I'll go thru @boblail comments and do code changes

@boblail done, feel free to take another look. I have left few comments to code to help better understand some place which are still not ideal.

> 2. I'll make through relation to raise for now. I can open my idea how to implement it in there as well, it still should be possible.

Bulk insert on now raise when called on through relation (current implementation actually leads to wrong inserts). It is possible to support bulk insert on through relation, but it will need some changes and will be separated by adapter. I think this will be great follow up. Also currently discussed pull request with `WITH` clause support will help implementing that (at least for some adapters).

I can also separate this change to custom PR since I think this should be merged A$AP. insert_all on through relation leads to unexpected results. Just ping me if you agree on this.

**Review by boblail — APPROVED** [review]
This looks good to me! 😄

**Comment by simi** [comment]
:cry:

**Comment by simi** [comment]
I'll focus on maintain and support activerecord-import gem instead of this. It doesn't seems to have enough attention and currently this feature is only half-baked in (not really useful for our cases). Thanks everyone for time spend on this.

**Comment by kaspth** [comment]
After using `insert_all` I've really wanted this too but I always feel quite out of my depth for reviewing Active Record PRs. @kamipo how does this look to you with a quick cursory glance?

**Comment by kamipo** [comment]
Use `model.scope_attributes` rather than `scope_for_create` directly like #38899.

**Comment by kamipo** [comment]
Closing in favor of #38899.

**Pull request closed without merging**
